### DISCUSSION TITLE
Workaround for Gnome Text Edit, Chromium, etc on Wayland?

### DISCUSSION THREAD

**Discussion by HubKing** [fcitx/fcitx5#693] [Q&A] [answered]
It seems that FCITX5 does not work on [Gnome Text Editor](https://gitlab.gnome.org/GNOME/gnome-text-editor) and all Chromium-based browsers if Wayland is enabled by setting `Preferred Ozone platform` to Auto in `chrome://flags`. The DE is Gnome 43.2, and I am using Wayland.

Is there any workaround?

**Comment by wengxt** [comment, marked as answer]
as for chrome ,you need --gtk-version=4 flag to make it able to use gtk4 im module.

For Gnome text editor it should work. What's your distro and how did you install gnome-text-editor? Did your distro split fcitx5-gtk in to a separate gtk4 package?

**Reply by HubKing** [reply]
I had tried `--gtk-version=4`  but Fcitx5 still does not work (only Chromium's theme becomes different). I install the text editor from the official repository.

![image](https://user-images.githubusercontent.com/10900489/209477717-5168fa07-3fb5-4286-94f0-8ed07a2312bc.png)

It is Gnome Manjaro. 

Yeah, I think so. It has a bunch of "fcitx-" and a bunch of "fcitx5-". I have installed the latter.

![image](https://user-images.githubusercontent.com/10900489/209477770-8752251e-cb2f-4f71-a774-abfad077899c.png)

**Reply by HubKing** [reply]
It seems that not only on the text editor, but Fcitx is not working on any GTK4 app.

**Reply by wengxt** [reply]
Then you might didn't set GTK_IM_MODULE?

Can you paste the output of fcitx5-diagnose?

For chromium this works for me
```
GTK_IM_MODULE=fcitx chromium --enable-features=UseOzonePlatform --ozone-platform=wayland --gtk-version=4
```

**Reply by HubKing** [reply]
`GTK_IM_MODULE=fcitx` made no difference.

Here is the output of diagnostics: [output.txt](https://github.com/fcitx/fcitx5/files/10300520/output.txt)

**Reply by wengxt** [reply]
I saw gnome-text-editor is already connected to fcitx, so I think it should already work?

          IC [fc402107f14245bba3ba8b3978d797fb] program:gnome-text-editor frontend:dbus cap:4001000012 focus:0

Or it could be a hotkey conflict? Does "Ctrl + ;" show the fcitx clipboard in gnome-text-editor?

As for chromium, how did you install it?

I tried with chromium arch repo and from AUR both seems to be work for me.

**Reply by HubKing** [reply]
Chromium was from the official repository.

Ctrl+; worked. I was using "Windows + Space" as the only shortcut for changing language, so I added a weird combination like Ctrl+shift+I in addition, and that worked on both the text editor and Chromium. I do not know why Windows + Space works on GTK3 apps and other apps like FireFox, but not on GTK4 and Chromium.

**Comment by HubKing** [comment]
I did some more testing and it seems that Fcitx cannot get any combination of Windows key (or "Super" as Gnome calls it) shortcut on any GTK4 application. It can, however, get such shortcuts on GTK3/QT applications.

Is this a new limitation or a bug of GTK4, or is it because Fcitx is not handling the shortcuts correctly on GTK4 applications? If it is the former, I may report this to GTK developers.

**Reply by wengxt** [reply]
Yes, you're right. Seems gtk doesn't report native xkb modifier state anymore, but only it's virtual super key state, which fcitx doesn't recognize as super. Need to make some change to make it work.

**Reply by wengxt** [reply]
> Yes, you're right. Seems gtk doesn't report native xkb modifier state anymore, but only it's virtual super key state, which fcitx doesn't recognize as super. Need to make some change to make it work.

Sorry I seems to misunderstood the actual issue, yeah, the modifier state of super is not correct in gtk4 ... let me check closer..

**Reply by wengxt** [reply]
Ok, two separate issues. Under gnome is exactly what I mentioned earlier. under kwin it's unclear to me, might be a gtk4 bug.

**Reply by wengxt** [reply]
Reported: https://gitlab.gnome.org/GNOME/gtk/-/issues/5476

Maybe you want to try ctrl+space as a workaround. (Personally I'd say "single right control" is another good candidate)

**Reply by wengxt** [reply]
Ah, I forget you're using gnome, 

So with https://github.com/fcitx/fcitx5/commit/469e849fc0bbd104b7a86b4b431fa6b6c4f30ad5 it should work under gnome now.

**Reply by HubKing** [reply]
Thanks. So, it means that, in future releases, Win+Space would work for GTK4 apps with https://github.com/fcitx/fcitx5/commit/469e849fc0bbd104b7a86b4b431fa6b6c4f30ad5 ? The default key was Ctrl+Space, I think, but that key combination is used by many code editors to show auto-complete, so I had been using Win+Space. Alt+Space is used by the window menu, so I have added Shift+Space temporarily for GTK4 apps.
